ntp-proto/src/nts_record.rs

```diff
@@ -905,27 +919,54 @@ impl KeyExchangeResultDecoder {
                     .iter()
                     .find_map(|proto| protocol_ids.contains(&(*proto as u16)).then_some(*proto));
 
-                state.protocol = selected;
-
-                match state.protocol {
+                match selected {
                     None => Break(Err(NoValidProtocol)),
-                    Some(_) => Continue(state),
+                    Some(protocol) => {
+                        // The NTS Next Protocol Negotiation record [..] MUST occur exactly once in every NTS-KE request and response.
+                        match state.protocol {
+                            None => {
+                                state.protocol = Some(protocol);
+                                Continue(state)
+                            }
+                            Some(_) => Break(Err(KeyExchangeError::BadRequest)),
```

Comment:
BadRequest seems a weird description for an error which is essentially "server did not give valid response"